supervisor/supervisor.go

```diff
@@ -0,0 +1,343 @@
+
+	if err == nil {
+		level.Info(s.logger).Log("msg", "sidecar supervisor exiting")
+		return nil
+	}
+	return err
+}
+
+func (s *Supervisor) supervise(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	for {
+		sleep := config.DefaultSupervisorPeriod
+		ready, _ := s.checkTarget()
+		if !ready {
+			sleep = config.DefaultHealthCheckTimeout
+		}
+
+		// Sleep or be canceled.
+		if func() bool {
+			tick := time.NewTicker(sleep)
+			defer tick.Stop()
+			select {
+			case <-ctx.Done():
+				return true
+			case <-tick.C:
+				return false
+			}
+		}() {
+			return
+		}
+
+		if err := s.healthcheck(ctx); err != nil {
+			if err != context.Canceled {
+				level.Error(s.logger).Log("msg", "healthcheck failed", "err", err)
+			}
+		}
+	}
+}
+
+func (s *Supervisor) healthcheck(ctx context.Context) (err error) {
+	ctx, cancel := context.WithTimeout(ctx, config.DefaultHealthCheckTimeout)
+	defer cancel()
+
+	ready, target := s.checkTarget()
+
+	ctx, span := tracer.Start(ctx, "health-client")
+
+	defer func() {
+		if err != nil {
+			span.RecordError(err)
+		}
+		span.End()
+	}()
+
+	req, err := http.NewRequestWithContext(ctx, "GET", target, nil)
+	if err != nil {
+		return errors.Wrap(err, "build request")
+	}
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "healtcheck GET")
+	}
+	defer resp.Body.Close()
+
+	var hstat string
+	if ready {
+		var hr health.Response
+		if err := json.NewDecoder(resp.Body).Decode(&hr); err != nil {
+			return errors.Wrap(err, "decode response")
+		}
+		span.SetAttributes(label.String("sidecar.status", hr.Status))
+
+		hstat = "ok"
+	} else if resp.StatusCode/100 == 2 {
+		s.noteReady()
+		hstat = "first time ready"
+		level.Info(s.logger).Log("msg", "sidecar is ready")
+	} else {
+		hstat = "not ready yet"
+		level.Info(s.logger).Log("msg", "sidecar is not ready")
+	}
+
+	span.AddEvent("recent-logs", trace.WithAttributes(
+		label.Array("activity", s.copyLogs()),
+	))
+
+	span.SetAttributes(label.String("sidecar.health", hstat))
+	span.SetAttributes(semconv.HTTPAttributesFromHTTPStatusCode(resp.StatusCode)...)
+	span.SetStatus(semconv.SpanStatusFromHTTPStatusCode(resp.StatusCode))
+	return nil
+}
+
+func (s *Supervisor) finalSpan(err error) {
+	_, span := tracer.Start(context.Background(), "shutdown-report")
+	defer span.End()
+
+	span.AddEvent("recent-logs", trace.WithAttributes(
+		label.Array("activity", s.copyLogs()),
+	))
+
+	if err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, "unexpected shutdown")
+		return
+	}
+	span.SetStatus(codes.Ok, "graceful shutdown")
+}
+
+func (s *Supervisor) newPipe(ctx context.Context, name string, real io.Writer, wg *sync.WaitGroup) *io.PipeWriter {
+	wg.Add(1)
+
+	rd, wr := io.Pipe()
+
+	go s.readOutput(ctx, name, real, rd, wg)
+
+	return wr
+}
+
+func (s *Supervisor) readOutput(_ context.Context, name string, real io.Writer, rd *io.PipeReader, wg *sync.WaitGroup) {
```

Comment:
This is a bit tricky, but the channels and contexts and defer statements are arranged so that, if the sidecar crashes or exits with a stack dump, the final span will report complete information.